File: src/conductor/candidate_review/identity.py
"""One identity for claiming a path and for writing it.

The claim store is only a control if the name a lane claims under is the name its
write gate checks. It was not. ``.codex/hooks.json`` pinned ``GOVERNANCE_OWNER=codex``
while codex sessions created claims as ``codex-audit-repository-scan-rust-20260903``,
so codex's gate could not recognise codex's own claims; every Claude session resolved
to the bare literal ``claude``, so any one of them could write over another's claim.
Both ends now call :func:`resolve_owner`.

The name is the **lane**, not the vendor. A lane is one worktree carrying one piece of
work, which is exactly the unit claims are meant to separate -- four codex worktrees
are four lanes, and calling them all ``codex`` is the same as not claiming. A vendor
literal is therefore not an identity here: it is ignored wherever it appears, including
in an exported ``GOVERNANCE_OWNER``, so a lane inherits a real name without every
launcher having to be edited first.

Derivation is filesystem-only. This runs on every Edit, Write and Bash in the fleet;
a ``git rev-parse`` per call would put a subprocess in the hook's latency path.
"""

from __future__ import annotations

import os
import re
from pathlib import Path
from typing import Mapping

# Vendor names are what the fleet used before lanes had identities. They stay
# resolvable (a lane with no derivable name still gets *something*, and the gate
# still honours claims written under them) but they are never a lane identity.
VENDOR_MARKERS: tuple[tuple[str, str], ...] = (
    ("QWEN_PROJECT_DIR", "qwen"),
    ("GROK_PROJECT_DIR", "grok"),
    ("CLAUDE_PROJECT_DIR", "claude"),
    ("CODEX_HOME", "codex"),
)
VENDORS: frozenset[str] = frozenset(vendor for _marker, vendor in VENDOR_MARKERS)
_OWNER_MAX = 64
_UNSAFE = re.compile(r"[^a-z0-9._-]+")


class OwnerIdentityError(RuntimeError):
    """The running lane cannot be named, so nothing may be claimed or written."""


def is_vendor(owner: str) -> bool:
    return owner.strip().casefold() in VENDORS


def vendor_of(env: Mapping[str, str] | None = None) -> str:
    """The vendor whose launcher this process is under, or ``""``."""
    env = os.environ if env is None else env
    for marker, vendor in VENDOR_MARKERS:
        if (env.get(marker) or "").strip():
            return vendor
    return ""


def vendor_for(owner: str, env: Mapping[str, str] | None = None) -> str:
    """The vendor a lane belongs to: its launcher's, else its own name prefix.

    The launcher marker is absent whenever the gate is reached from a plain shell,
    so the lane name has to be able to answer on its own. Lanes are named
    ``<vendor>-<work>-<date>`` by the worktree convention, which makes the prefix
    a reliable second source and keeps the legacy-claim fallback from silently
    going dark in exactly the sessions that still need it.
    """
    marker = vendor_of(env)
    if marker:
        return marker
    head = normalize(owner).split("-", 1)[0]
    return head if head in VENDORS else ""


def normalize(owner: str) -> str:
    """Fold an arbitrary lane string into the owner charset, or ``""``."""
    folded = _UNSAFE.sub("-", owner.strip().casefold()).strip("-.")
    return folded[:_OWNER_MAX]


def lane_of(repo_root: Path) -> str:
    """The lane name of a checkout: its worktree name, else its branch.

    A linked worktree is named for the work it carries, and keeps that name across
    branch switches, so it is the better key. The main checkout is named ``LLM``,
    which is shared by every session in it, so there the branch is the lane.
    """
    marker = repo_root / ".git"
    if marker.is_file():
        return normalize(repo_root.name)
    if not marker.is_dir():
        return ""
    try:
        head = (marker / "HEAD").read_text(encoding="utf-8").strip()
    except OSError:
        return ""
    if not head.startswith("ref: refs/heads/"):
        return ""  # detached: no branch, so no lane
    return normalize(head[len("ref: refs/heads/") :])


def resolve_owner(
    repo_root: Path | None = None, env: Mapping[str, str] | None = None
) -> str:
    """The identity this lane claims and writes under.

    An exported ``GOVERNANCE_OWNER`` wins -- that is a lane declaring its own name --
    unless it is a bare vendor literal, which is what the old launchers pinned and
    what this exists to stop. Otherwise the lane is derived from the checkout, and
    only if that fails does the vendor stand in.
    """
    env = os.environ if env is None else env
    declared = normalize(env.get("GOVERNANCE_OWNER") or "")
    if declared and not is_vendor(declared):
        return declared
    if repo_root is not None:
        lane = lane_of(repo_root)
        if lane and not is_vendor(lane):
            return lane
    vendor = vendor_of(env) or declared
    if vendor:
        return vendor
    raise OwnerIdentityError(
        "no governance identity: this process is under no known agent launcher and "
        "its checkout has no lane name — export GOVERNANCE_OWNER=<lane> "
        "(a worktree or branch name, not a vendor)"
    )


def require_lane_owner(owner: str) -> str:
    """Validate an owner about to be written into a claim."""
    folded = normalize(owner)
    if not folded:
        raise OwnerIdentityError(f"owner {owner!r} is empty after normalization")
    if is_vendor(folded):
        raise OwnerIdentityError(
            f"owner {folded!r} names a vendor, not a lane: every session of that "
            "vendor would satisfy this claim. Claim under the worktree or branch "
            "name, or export GOVERNANCE_OWNER=<lane>"
        )
    return folded

File: native/forge/src/identity.rs
//! Native port of `conductor.candidate_review.identity`: one identity for
//! claiming a path and for writing it. See the Python module's docstring for
//! the full "codex claimed as X but denied itself" motivation -- this port
//! preserves every rule verbatim, including the vendor-is-not-a-lane
//! distinction and the derive-then-fall-back-to-vendor order in
//! `resolve_owner`.
//!
//! Derivation is filesystem-only (no `git` subprocess), matching the Python
//! module's own stated design ("a `git rev-parse` per call would put a
//! subprocess in the hook's latency path").

use std::collections::HashMap;
use std::path::Path;

/// `VENDOR_MARKERS`: env var -> vendor name, checked in this order.
const VENDOR_MARKERS: &[(&str, &str)] = &[
    ("QWEN_PROJECT_DIR", "qwen"),
    ("GROK_PROJECT_DIR", "grok"),
    ("CLAUDE_PROJECT_DIR", "claude"),
    ("CODEX_HOME", "codex"),
];

const OWNER_MAX: usize = 64;

/// The running lane cannot be named, so nothing may be claimed or written.
#[derive(Debug)]
pub struct OwnerIdentityError(pub String);

impl std::fmt::Display for OwnerIdentityError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl std::error::Error for OwnerIdentityError {}

fn is_vendor(owner: &str) -> bool {
    let folded = owner.trim().to_lowercase();
    VENDOR_MARKERS.iter().any(|(_, vendor)| *vendor == folded)
}

/// `vendor_of`: the vendor whose launcher this process is under, or `""`.
fn vendor_of(env: &HashMap<String, String>) -> String {
    for (marker, vendor) in VENDOR_MARKERS {
        if env
            .get(*marker)
            .map(|v| !v.trim().is_empty())
            .unwrap_or(false)
        {
            return vendor.to_string();
        }
    }
    String::new()
}

/// `vendor_for`: the vendor a lane belongs to -- its launcher's, else its own
/// name prefix.
pub fn vendor_for(owner: &str, env: &HashMap<String, String>) -> String {
    let marker = vendor_of(env);
    if !marker.is_empty() {
        return marker;
    }
    let normalized = normalize(owner);
    let head = normalized.split('-').next().unwrap_or("");
    if VENDOR_MARKERS.iter().any(|(_, vendor)| *vendor == head) {
        head.to_string()
    } else {
        String::new()
    }
}

/// `normalize`: fold an arbitrary lane string into the owner charset, or `""`.
/// Mirrors Python's `re.sub(r"[^a-z0-9._-]+", "-", owner.strip().casefold())
/// .strip("-.")[:64]`.
pub fn normalize(owner: &str) -> String {
    let folded = owner.trim().to_lowercase();
    let mut out = String::with_capacity(folded.len());
    let mut in_run = false;
    for ch in folded.chars() {
        let safe = ch.is_ascii_alphanumeric() || matches!(ch, '.' | '_' | '-');
        if safe {
            out.push(ch);
            in_run = false;
        } else if !in_run {
            out.push('-');
            in_run = true;
        }
    }
    let trimmed = out.trim_matches(|c| c == '-' || c == '.');
    let truncated: String = trimmed.chars().take(OWNER_MAX).collect();
    truncated
}

/// `lane_of`: the lane name of a checkout -- its worktree name, else its
/// branch. Filesystem-only: reads `.git` (file for a linked worktree,
/// directory for the main checkout) and, for the latter, `.git/HEAD`.
pub fn lane_of(repo_root: &Path) -> String {
    let marker = repo_root.join(".git");
    if marker.is_file() {
        let name = repo_root.file_name().and_then(|n| n.to_str()).unwrap_or("");
        return normalize(name);
    }
    if !marker.is_dir() {
        return String::new();
    }
    let Ok(head) = std::fs::read_to_string(marker.join("HEAD")) else {
        return String::new();
    };
    let head = head.trim();
    let Some(branch) = head.strip_prefix("ref: refs/heads/") else {
        return String::new(); // detached: no branch, so no lane
    };
    normalize(branch)
}

/// `resolve_owner`: the identity this lane claims and writes under. `Err`
/// when no identity can be derived (mirrors `OwnerIdentityError`).
pub fn resolve_owner(
    repo_root: Option<&Path>,
    env: &HashMap<String, String>,
) -> Result<String, OwnerIdentityError> {
    let declared = normalize(
        env.get("GOVERNANCE_OWNER")
            .map(String::as_str)
            .unwrap_or(""),
    );
    if !declared.is_empty() && !is_vendor(&declared) {
        return Ok(declared);
    }
    if let Some(root) = repo_root {
        let lane = lane_of(root);
        if !lane.is_empty() && !is_vendor(&lane) {
            return Ok(lane);
        }
    }
    let vendor = {
        let v = vendor_of(env);
        if v.is_empty() {
            declared
        } else {
            v
        }
    };
    if !vendor.is_empty() {
        return Ok(vendor);
    }
    Err(OwnerIdentityError(
        "no governance identity: this process is under no known agent launcher and \
         its checkout has no lane name — export GOVERNANCE_OWNER=<lane> \
         (a worktree or branch name, not a vendor)"
            .to_string(),
    ))
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::atomic::{AtomicU32, Ordering};

    fn env(pairs: &[(&str, &str)]) -> HashMap<String, String> {
        pairs
            .iter()
            .map(|(k, v)| (k.to_string(), v.to_string()))
            .collect()
    }

    struct ScratchRepo(std::path::PathBuf);

    impl ScratchRepo {
        fn new(label: &str) -> Self {
            static COUNTER: AtomicU32 = AtomicU32::new(0);
            let n = COUNTER.fetch_add(1, Ordering::Relaxed);
            let dir = std::env::temp_dir().join(format!(
                "forge-identity-test-{}-{label}-{n}",
                std::process::id()
            ));
            std::fs::create_dir_all(&dir).unwrap();
            ScratchRepo(dir)
        }
    }

    impl Drop for ScratchRepo {
        fn drop(&mut self) {
            let _ = std::fs::remove_dir_all(&self.0);
        }
    }

    #[test]
    fn normalize_folds_unsafe_characters_and_casefolds() {
        assert_eq!(normalize("  Codex Audit/Scan!! "), "codex-audit-scan");
        assert_eq!(normalize("--.trim.me.--"), "trim.me");
        assert_eq!(normalize(""), "");
    }

    #[test]
    fn normalize_truncates_to_64_bytes() {
        let long = "a".repeat(100);
        assert_eq!(normalize(&long).len(), 64);
    }

    #[test]
    fn is_vendor_matches_known_vendors_case_insensitively() {
        assert!(is_vendor("Codex"));
        assert!(is_vendor(" claude "));
        assert!(!is_vendor("codex-audit"));
    }

    #[test]
    fn vendor_of_checks_markers_in_declared_order() {
        let e = env(&[("CLAUDE_PROJECT_DIR", "/x"), ("CODEX_HOME", "/y")]);
        assert_eq!(vendor_of(&e), "claude");
        assert_eq!(vendor_of(&HashMap::new()), "");
    }

    #[test]
    fn vendor_for_prefers_the_launcher_marker_over_the_name_prefix() {
        let e = env(&[("CODEX_HOME", "/y")]);
        assert_eq!(vendor_for("claude-something", &e), "codex");
    }

    #[test]
    fn vendor_for_falls_back_to_a_recognized_name_prefix() {
        assert_eq!(vendor_for("codex-audit-scan", &HashMap::new()), "codex");
        assert_eq!(vendor_for("llm-b0-lane", &HashMap::new()), "");
    }

    #[test]
    fn lane_of_reads_the_worktree_name_from_a_linked_worktree_git_file() {
        let repo = ScratchRepo::new("worktree");
        let named = repo.0.join("llm-b0-my-lane");
        std::fs::create_dir_all(&named).unwrap();
        std::fs::write(named.join(".git"), "gitdir: /elsewhere/.git\n").unwrap();
        assert_eq!(lane_of(&named), "llm-b0-my-lane");
    }

    #[test]
    fn lane_of_reads_the_branch_from_a_main_checkout() {
        let repo = ScratchRepo::new("main");
        std::fs::create_dir_all(repo.0.join(".git")).unwrap();
        std::fs::write(
            repo.0.join(".git/HEAD"),
            "ref: refs/heads/forge/my-branch\n",
        )
        .unwrap();
        assert_eq!(lane_of(&repo.0), "forge-my-branch");
    }

    #[test]
    fn lane_of_is_empty_for_a_detached_head() {
        let repo = ScratchRepo::new("detached");
        std::fs::create_dir_all(repo.0.join(".git")).unwrap();
        std::fs::write(repo.0.join(".git/HEAD"), "abcdef0123456789\n").unwrap();
        assert_eq!(lane_of(&repo.0), "");
    }

    #[test]
    fn lane_of_is_empty_outside_any_checkout() {
        let repo = ScratchRepo::new("none");
        assert_eq!(lane_of(&repo.0), "");
    }

    #[test]
    fn resolve_owner_prefers_a_declared_non_vendor_governance_owner() {
        let e = env(&[("GOVERNANCE_OWNER", "llm-b0")]);
        assert_eq!(resolve_owner(None, &e).unwrap(), "llm-b0");
    }

    #[test]
    fn resolve_owner_ignores_a_declared_vendor_literal() {
        let repo = ScratchRepo::new("declared-vendor");
        std::fs::create_dir_all(repo.0.join(".git")).unwrap();
        std::fs::write(repo.0.join(".git/HEAD"), "ref: refs/heads/my-lane\n").unwrap();
        let e = env(&[("GOVERNANCE_OWNER", "codex")]);
        assert_eq!(resolve_owner(Some(&repo.0), &e).unwrap(), "my-lane");
    }

    #[test]
    fn resolve_owner_falls_back_to_the_checkout_lane() {
        let repo = ScratchRepo::new("lane-fallback");
        std::fs::create_dir_all(repo.0.join(".git")).unwrap();
        std::fs::write(repo.0.join(".git/HEAD"), "ref: refs/heads/my-lane\n").unwrap();
        assert_eq!(
            resolve_owner(Some(&repo.0), &HashMap::new()).unwrap(),
            "my-lane"
        );
    }

    #[test]
    fn resolve_owner_falls_back_to_the_vendor_marker_when_the_lane_is_unnameable() {
        let repo = ScratchRepo::new("vendor-fallback");
        // Detached HEAD: no lane.
        std::fs::create_dir_all(repo.0.join(".git")).unwrap();
        std::fs::write(repo.0.join(".git/HEAD"), "abcdef\n").unwrap();
        let e = env(&[("CLAUDE_PROJECT_DIR", "/x")]);
        assert_eq!(resolve_owner(Some(&repo.0), &e).unwrap(), "claude");
    }

    #[test]
    fn resolve_owner_errors_when_nothing_can_name_the_lane() {
        let repo = ScratchRepo::new("unnameable");
        assert!(resolve_owner(Some(&repo.0), &HashMap::new()).is_err());
        assert!(resolve_owner(None, &HashMap::new()).is_err());
    }
}
